kaiax/builder/impl/tx_pool_wrapping.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2025 The Kaia Authors
+// This file is part of the Kaia library.
+//
+// The Kaia library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Kaia library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Kaia library. If not, see <http://www.gnu.org/licenses/>.
+
+package impl
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/common"
+	"github.com/kaiachain/kaia/kaiax"
+	"github.com/kaiachain/kaia/kaiax/builder"
+)
+
+var _ kaiax.TxPoolModule = (*BuilderWrappingModule)(nil)
+
+type BuilderWrappingModule struct {
+	txBundlingModule builder.TxBundlingModule
+	txPoolModule     kaiax.TxPoolModule // either nil or same object as txBundlingModule
+	knownTxs         map[common.Hash]txAndTime
+	mu               sync.RWMutex
+}
+
+func NewBuilderWrappingModule(txBundlingModule builder.TxBundlingModule) *BuilderWrappingModule {
+	txPoolModule, _ := txBundlingModule.(kaiax.TxPoolModule)
+	return &BuilderWrappingModule{
+		txBundlingModule: txBundlingModule,
+		txPoolModule:     txPoolModule,
+		knownTxs:         make(map[common.Hash]txAndTime),
+	}
+}
+
+// PreAddTx is a wrapper function that calls the PreAddTx method of the underlying module.
+func (b *BuilderWrappingModule) PreAddTx(tx *types.Transaction, local bool) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
```

Comment:
mutex for `knownTxs`, right?

```suggestion
	b.mu.RLock()
	defer b.mu.RUnlock()
```

Reply:
yes. fixed at https://github.com/kaiachain/kaia/pull/309/commits/9f2419e6b943d82c29c7aa32cafd1242cd89845c.